core/src/main/scala/org/bitcoins/core/wallet/signer/Signer.scala

```diff
@@ -119,271 +130,248 @@ sealed abstract class Signer {
-                       signers(i),
-                       hashType,
-                       isDummySignature))
-          val signatures = Future.sequence(signaturesNested)
-          signatures.map { sigs =>
-            val multiSigScriptSig = MultiSignatureScriptSignature(sigs)
-            val signedInput = TransactionInput(unsignedInput.previousOutput,
-                                               multiSigScriptSig,
-                                               unsignedInput.sequence)
-            val signedInputs =
-              unsignedTx.inputs.updated(inputIndex.toInt, signedInput)
-            val signedTx = unsignedTx match {
-              case btx: BaseTransaction =>
-                BaseTransaction(btx.version,
-                                signedInputs,
-                                btx.outputs,
-                                btx.lockTime)
-              case wtx: WitnessTransaction =>
-                WitnessTransaction(wtx.version,
-                                   signedInputs,
-                                   wtx.outputs,
-                                   wtx.lockTime,
-                                   wtx.witness)
-            }
-            BaseTxSigComponent(signedTx, inputIndex, output, flags)
-          }
-        }
-      case _: P2PKScriptPubKey | _: P2PKHScriptPubKey | _: P2SHScriptPubKey |
-          _: P2WPKHWitnessSPKV0 | _: P2WSHWitnessSPKV0 |
-          _: LockTimeScriptPubKey | _: NonStandardScriptPubKey |
-          _: WitnessCommitment | _: UnassignedWitnessScriptPubKey |
-          EmptyScriptPubKey =>
-        Future.fromTry(TxBuilderError.WrongSigner)
+
+    val requiredSigs = spendingInfoToSatisfy.scriptPubKey.requiredSigs
+    val signatureFs = 0
+      .until(requiredSigs)
+      .map(
+        i =>
+          doSign(sigComponent(spendingInfo, unsignedTx),
+                 signers(i),
+                 hashType,
+                 isDummySignature))
+
+    val signaturesF = Future.sequence(signatureFs)
+
+    val scriptSigF = signaturesF.map { sigs =>
+      MultiSignatureScriptSignature(sigs)
     }
-    signed
+
+    updateScriptSigInSigComponent(unsignedTx,
+                                  inputIndex.toInt,
+                                  output,
+                                  scriptSigF)
   }
 }
 
 object MultiSigSigner extends MultiSigSigner
 
-sealed abstract class P2WPKHSigner extends BitcoinSigner {
+sealed abstract class P2WPKHSigner extends BitcoinSigner[P2WPKHV0SpendingInfo] {
 
   override def sign(
       spendingInfo: UTXOSpendingInfo,
       unsignedTx: Transaction,
       isDummySignature: Boolean,
-      scriptPubKeyToSatisfy: ScriptPubKey)(
+      spendingInfoToSatisfy: P2WPKHV0SpendingInfo)(
       implicit ec: ExecutionContext): Future[TxSigComponent] = {
-    if (scriptPubKeyToSatisfy != spendingInfo.output.scriptPubKey) {
+    if (spendingInfoToSatisfy != spendingInfo) {
       Future.fromTry(TxBuilderError.NestedWitnessSPK)
```

Comment:
is this this right error here still? Shouldn't it be like `TxBuilderError.WrongSigner`?

Reply:
yeah, `spendingInfoToSatisfy != spendingInfo` => Nesting

Reply:
How do you know that `spendingInfo` is a witness (specifically `SegwitV0NativeUTXOSpendingInfo`)? The error says `nestedWitnessSPK`

Reply:
If they weren't equal then that means this is a nested call that was made via a match on the nested info and this is a witness signer